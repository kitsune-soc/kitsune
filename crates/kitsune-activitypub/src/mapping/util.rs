use super::State;
use kitsune_db::model::{account::Account, post::Visibility};
use kitsune_type::ap::PUBLIC_IDENTIFIER;

pub trait BaseToCc {
    fn base_to_cc(&self, state: State<'_>, account: &Account) -> (Vec<String>, Vec<String>);
}

impl BaseToCc for Visibility {
    #[inline]
    fn base_to_cc(&self, state: State<'_>, account: &Account) -> (Vec<String>, Vec<String>) {
        let followers_url = state.service.url.followers_url(account.id);

        match self {
            Visibility::Public => (vec![PUBLIC_IDENTIFIER.to_string()], vec![followers_url]),
            Visibility::Unlisted => (vec![], vec![PUBLIC_IDENTIFIER.to_string(), followers_url]),
            Visibility::FollowerOnly => (vec![followers_url], vec![]),
            Visibility::MentionOnly => (vec![], vec![]),
        }
    }
}
