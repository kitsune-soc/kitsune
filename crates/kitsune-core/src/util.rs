use crate::state::State;
use hex_simd::{AsOut, AsciiCase};
use iso8601_timestamp::Timestamp;
use kitsune_db::model::{account::Account, oauth2::access_token::AccessToken, post::Visibility};
use kitsune_type::ap::PUBLIC_IDENTIFIER;
use pulldown_cmark::{html, Options, Parser};

const TOKEN_LENGTH: usize = 32;

#[inline]
#[must_use]
pub fn generate_secret() -> String {
    let token_data: [u8; TOKEN_LENGTH] = rand::random();
    let mut buf = [0_u8; TOKEN_LENGTH * 2];
    (*hex_simd::encode_as_str(&token_data, buf.as_mut_slice().as_out(), AsciiCase::Lower))
        .to_string()
}

#[inline]
#[must_use]
pub fn process_markdown(markdown: &str) -> String {
    let parser = Parser::new_ext(markdown, Options::all());
    let mut buf = String::new();
    html::push_html(&mut buf, parser);
    buf
}

pub trait AccessTokenTtl {
    fn ttl(&self) -> time::Duration;
}

impl AccessTokenTtl for AccessToken {
    #[inline]
    fn ttl(&self) -> time::Duration {
        *self.expires_at - *Timestamp::now_utc()
    }
}

pub trait BaseToCc {
    fn base_to_cc(&self, state: &State, account: &Account) -> (Vec<String>, Vec<String>);
}

impl BaseToCc for Visibility {
    #[inline]
    fn base_to_cc(&self, state: &State, account: &Account) -> (Vec<String>, Vec<String>) {
        let followers_url = state.service.url.followers_url(account.id);

        match self {
            Visibility::Public => (vec![PUBLIC_IDENTIFIER.to_string()], vec![followers_url]),
            Visibility::Unlisted => (vec![], vec![PUBLIC_IDENTIFIER.to_string(), followers_url]),
            Visibility::FollowerOnly => (vec![followers_url], vec![]),
            Visibility::MentionOnly => (vec![], vec![]),
        }
    }
}
