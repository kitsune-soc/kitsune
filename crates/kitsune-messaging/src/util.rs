use std::fmt;

pub struct TransparentDebug<T>(pub T);

impl<T> fmt::Debug for TransparentDebug<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} {{ ... }}", std::any::type_name::<T>())
    }
}
