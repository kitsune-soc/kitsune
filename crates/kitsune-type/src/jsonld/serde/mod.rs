//! Serde helpers to translate JSON-LD data structures.
//!
//! ## JSON-LD `@set`
//!
//! When a JSON-LD term's `@container` is unspecified or is set to `@set`, JSON entry values in the
//! following groups are semantically equivalent, respectively:
//!
//! - A non-array value (`"value"`) and a single-value array of the same value (`["value"]`)
//! - An empty array (`[]`), `null` and an absent entry
//!
//! The following helpers in the module deserialise a set as a sequence:
//!
//! - [`Set`]
//! - [`IdSet`]
//!
//! The following helpers deserialise a single value or `null` from a set:
//!
//! - [`First`]
//! - [`FirstId`]
//! - [`FirstOk`]
//!
//! ## JSON-LD `@id`
//!
//! When a JSON-LD term's `@type` is set to `@id`, a JSON entry value of a single (IRI) string
//! (`"http://example.com/"`) is a shorthand for a Linked Data node identified by that string
//! (`{"@id": "http://example.com/"}`.
//!
//! The following helpers deserialise the node identifier string(s):
//!
//! - [`Id`]
//! - [`FirstId`]
//! - [`IdSet`]

macro_rules! forward_to_into_deserializer {
    (
        fn visit_borrowed_str($T:ty);
        $($rest:tt)*
    ) => {
        fn visit_borrowed_str<E>(self, v: $T) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            self.0.deserialize(serde::de::value::BorrowedStrDeserializer::new(v))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_borrowed_bytes($T:ty);
        $($rest:tt)*
    ) => {
        fn visit_borrowed_bytes<E>(self, v: $T) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            self.0.deserialize(serde::de::value::BorrowedBytesDeserializer::new(v))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_none();
        $($rest:tt)*
    ) => {
        fn visit_none<E>(self) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            self.0.deserialize(serde::de::IntoDeserializer::into_deserializer(()))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_some();
        $($rest:tt)*
    ) => {
        fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
        where
            D: serde::Deserializer<'de>,
        {
            self.0.deserialize(deserializer)
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_unit();
        $($rest:tt)*
    ) => {
        fn visit_unit<E>(self) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            self.0.deserialize(serde::de::IntoDeserializer::into_deserializer(()))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_newtype_struct();
        $($rest:tt)*
    ) => {
        fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
        where
            D: serde::Deserializer<'de>,
        {
            self.0.deserialize(deserializer)
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_seq();
        $($rest:tt)*
    ) => {
        fn visit_seq<A>(self, seq: A) -> Result<Self::Value, A::Error>
        where
            A: serde::de::SeqAccess<'de>,
        {
            self.0.deserialize(serde::de::value::SeqAccessDeserializer::new(seq))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_map();
        $($rest:tt)*
    ) => {
        fn visit_map<A>(self, map: A) -> Result<Self::Value, A::Error>
        where
            A: serde::de::MapAccess<'de>,
        {
            self.0.deserialize(serde::de::value::MapAccessDeserializer::new(map))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn visit_enum();
        $($rest:tt)*
    ) => {
        fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
        where
            A: serde::de::EnumAccess<'de>,
        {
            self.0.deserialize(serde::de::value::EnumAccessDeserializer::new(data))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    (
        fn $name:ident($T:ty);
        $($rest:tt)*
    ) => {
        fn $name<E>(self, v: $T) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            self.0.deserialize(serde::de::IntoDeserializer::into_deserializer(v))
        }
        forward_to_into_deserializer! { $($rest)* }
    };
    () => {};
}

mod first;
mod first_id;
mod first_ok;
mod id;
mod id_set;
mod optional;
mod set;

pub use self::first::First;
pub use self::first_id::FirstId;
pub use self::first_ok::FirstOk;
pub use self::id::Id;
pub use self::id_set::IdSet;
pub use self::optional::Optional;
pub use self::set::Set;

use core::{
    fmt::{self, Formatter},
    marker::PhantomData,
};
use serde::de::{
    self,
    value::{EnumAccessDeserializer, MapAccessDeserializer, SeqAccessDeserializer},
    DeserializeSeed, Deserializer, EnumAccess, IntoDeserializer, MapAccess, SeqAccess, Visitor,
};

const EXPECTING_SET: &str = "a JSON-LD set";

/// A wrapper to implement `IntoDeserializer` for an `impl Deserializer`, because Serde somehow
/// doesn't provide a blanket impl.
struct DeserializerIntoDeserializer<D>(D);

/// A `DeserializeSeed` that catches a recoverable error and returns it as a successful value.
struct CatchError<T, E> {
    seed: T,
    marker: PhantomData<fn() -> E>,
}

struct OptionSeed<T>(Option<T>);

impl<'de, D> IntoDeserializer<'de, D::Error> for DeserializerIntoDeserializer<D>
where
    D: Deserializer<'de>,
{
    type Deserializer = D;

    fn into_deserializer(self) -> Self::Deserializer {
        self.0
    }
}

impl<'de, T> DeserializeSeed<'de> for &mut OptionSeed<T>
where
    T: DeserializeSeed<'de>,
{
    type Value = Option<T::Value>;

    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        if let Some(seed) = self.0.take() {
            seed.deserialize(deserializer).map(Some)
        } else {
            Ok(None)
        }
    }
}

impl<'de, T, E> CatchError<T, E>
where
    T: DeserializeSeed<'de>,
    E: de::Error,
{
    pub fn new(seed: T) -> Self {
        Self {
            seed,
            marker: PhantomData,
        }
    }
}

impl<'de, T, E> DeserializeSeed<'de> for CatchError<T, E>
where
    T: DeserializeSeed<'de>,
    E: de::Error,
{
    type Value = Result<T::Value, E>;

    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_any(self)
    }
}

macro_rules! catch_error_forward_to_into_deserializer {
    ($(fn $name:ident($T:ty);)*) => {$(
        fn $name<E2>(self, v: $T) -> Result<Self::Value, E2> {
            // We can tell that the error isn't fatal to the deserialiser because it's originated
            // from the already deserialised value `$t` rather than the deserialiser.
            Ok(self.seed.deserialize(serde::de::IntoDeserializer::into_deserializer(v)))
        }
    )*};
}

impl<'de, T, E> Visitor<'de> for CatchError<T, E>
where
    T: DeserializeSeed<'de>,
    E: de::Error,
{
    type Value = Result<T::Value, E>;

    fn expecting(&self, f: &mut Formatter<'_>) -> fmt::Result {
        f.write_str("a value")
    }

    fn visit_borrowed_str<E2>(self, v: &'de str) -> Result<Self::Value, E2>
    where
        E: de::Error,
    {
        Ok(self
            .seed
            .deserialize(de::value::BorrowedStrDeserializer::new(v)))
    }

    fn visit_borrowed_bytes<E2>(self, v: &'de [u8]) -> Result<Self::Value, E2>
    where
        E: de::Error,
    {
        Ok(self
            .seed
            .deserialize(de::value::BorrowedBytesDeserializer::new(v)))
    }

    fn visit_none<E2>(self) -> Result<Self::Value, E2> {
        Ok(self.seed.deserialize(().into_deserializer()))
    }

    fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        self.deserialize(deserializer)
    }

    fn visit_unit<E2>(self) -> Result<Self::Value, E2> {
        Ok(self.seed.deserialize(().into_deserializer()))
    }

    fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        self.deserialize(deserializer)
    }

    // XXX: The following methods cannot determine whether an error is recoverable. While we might
    // be able to implement them _right_ way by hooking into the `*Access` trait implementations and
    // recursively applying `CatchError` in the `*_seed` method calls, that wouldn't be worth the
    // effort since we're currently not using `FirstOk` (the only user of `CatchError` as of now)
    // for these types, and as for `visit_seq`, JSON-LD doesn't support nested sets anyway.
    fn visit_seq<A>(self, seq: A) -> Result<Self::Value, A::Error>
    where
        A: SeqAccess<'de>,
    {
        self.seed
            .deserialize(SeqAccessDeserializer::new(seq))
            .map(Ok)
    }

    fn visit_map<A>(self, map: A) -> Result<Self::Value, A::Error>
    where
        A: MapAccess<'de>,
    {
        self.seed
            .deserialize(MapAccessDeserializer::new(map))
            .map(Ok)
    }

    fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
    where
        A: EnumAccess<'de>,
    {
        self.seed
            .deserialize(EnumAccessDeserializer::new(data))
            .map(Ok)
    }

    catch_error_forward_to_into_deserializer! {
        fn visit_bool(bool);
        fn visit_i8(i8);
        fn visit_i16(i16);
        fn visit_i32(i32);
        fn visit_i64(i64);
        fn visit_i128(i128);
        fn visit_u8(u8);
        fn visit_u16(u16);
        fn visit_u32(u32);
        fn visit_u64(u64);
        fn visit_u128(u128);
        fn visit_f32(f32);
        fn visit_f64(f64);
        fn visit_char(char);
        fn visit_str(&str);
        fn visit_string(String);
        fn visit_bytes(&[u8]);
        fn visit_byte_buf(Vec<u8>);
    }
}

#[cfg(test)]
fn into_deserializer<'de, T>(value: T) -> T::Deserializer
where
    T: serde::de::IntoDeserializer<'de, serde::de::value::Error>,
{
    serde::de::IntoDeserializer::into_deserializer(value)
}

#[cfg(test)]
mod tests {
    use super::{First, FirstId, FirstOk, IdSet, Optional};
    use serde::Deserialize;

    /// Checks that the types work for some random real-world-ish use cases.
    #[test]
    fn integrate() {
        #[derive(Debug, Deserialize, PartialEq)]
        enum Type {
            Note,
        }

        #[derive(Debug, Deserialize, PartialEq)]
        #[serde(rename_all = "camelCase")]
        struct Object {
            id: String,
            #[serde(deserialize_with = "FirstOk::deserialize")]
            r#type: Type,
            #[serde(deserialize_with = "FirstId::deserialize")]
            attributed_to: String,
            #[serde(default)]
            #[serde(deserialize_with = "Optional::<First<_>>::deserialize")]
            summary: Option<String>,
            #[serde(default)]
            #[serde(deserialize_with = "Optional::<First<_>>::deserialize")]
            content: Option<String>,
            #[serde(default)]
            #[serde(deserialize_with = "IdSet::deserialize")]
            to: Vec<String>,
            #[serde(default)]
            #[serde(deserialize_with = "IdSet::deserialize")]
            cc: Vec<String>,
        }

        let object = br#"
        {
            "@context": "https://www.w3.org/ns/activitystreams",
            "id": "https://example.com/notes/1",
            "type": "Note",
            "attributedTo": "https://example.com/actors/1",
            "summary": "An ordinary Note",
            "content": "Hello, world!",
            "to": ["https://example.com/actors/2"],
            "cc": ["https://www.w3.org/ns/activitystreams#Public"]
        }
        "#;
        let expected = Object {
            id: "https://example.com/notes/1".to_owned(),
            r#type: Type::Note,
            attributed_to: "https://example.com/actors/1".to_owned(),
            summary: Some("An ordinary Note".to_owned()),
            content: Some("Hello, world!".to_owned()),
            to: vec!["https://example.com/actors/2".to_owned()],
            cc: vec!["https://www.w3.org/ns/activitystreams#Public".to_owned()],
        };
        assert_eq!(simd_json::from_slice(&mut object.to_owned()), Ok(expected));

        let object = br#"
        {
            "@context": "https://www.w3.org/ns/activitystreams",
            "id": "https://example.com/notes/1",
            "type": ["http://schema.org/CreativeWork", "Note"],
            "attributedTo": [
                {
                    "id": "https://example.com/actors/1",
                    "type": "Person"
                },
                "https://example.com/actors/2"
            ],
            "summary": "A quirky Note",
            "to": "https://example.com/actors/3"
        }
        "#;
        let expected = Object {
            id: "https://example.com/notes/1".to_owned(),
            // Multiple `type`s including unknown ones:
            r#type: Type::Note,
            // Multiple `attributedTo`s and an embedded node:
            attributed_to: "https://example.com/actors/1".to_owned(),
            summary: Some("A quirky Note".to_owned()),
            // Absent `Option` field:
            content: None,
            // Single-value set:
            to: vec!["https://example.com/actors/3".to_owned()],
            // Absent `serde(default)` field:
            cc: vec![],
        };
        assert_eq!(simd_json::from_slice(&mut object.to_owned()), Ok(expected));
    }
}
