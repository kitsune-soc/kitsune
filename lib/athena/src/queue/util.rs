use ahash::AHashMap;
use redis::{streams::StreamId, FromRedisValue, RedisResult};

#[derive(Clone, Debug)]
pub struct StreamAutoClaimReply {
    pub start_stream_id: String,
    pub claimed_ids: Vec<StreamId>,
    pub deleted_ids: Vec<String>,
}

impl FromRedisValue for StreamAutoClaimReply {
    fn from_redis_value(v: &redis::Value) -> RedisResult<Self> {
        type AutoClaimReturnValue = (
            String,
            Vec<Vec<(String, AHashMap<String, redis::Value>)>>,
            Vec<String>,
        );

        let (start_stream_id, claimed_ids, deleted_ids): AutoClaimReturnValue =
            redis::from_redis_value(v)?;

        let claimed_ids: Vec<StreamId> = claimed_ids
            .into_iter()
            .flat_map(|row| row.into_iter().map(|(id, map)| StreamId { id, map }))
            .collect();

        Ok(Self {
            start_stream_id,
            claimed_ids,
            deleted_ids,
        })
    }
}
