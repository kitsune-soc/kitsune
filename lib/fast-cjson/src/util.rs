use sonic_rs::writer::WriteExt;
use std::{fmt, io, mem::MaybeUninit};

macro_rules! for_both {
    ($owner:ident, $matcher:pat => $impl:expr) => {{
        match $owner {
            Self::Left($matcher) => $impl,
            Self::Right($matcher) => $impl,
        }
    }};
}

pub enum Either<L, R> {
    Left(L),
    Right(R),
}

impl<L, R> io::Write for Either<L, R>
where
    L: io::Write,
    R: io::Write,
{
    #[inline]
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        for_both!(self, inner => inner.write(buf))
    }

    #[inline]
    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
        for_both!(self, inner => inner.write_all(buf))
    }

    #[inline]
    fn write_fmt(&mut self, fmt: fmt::Arguments<'_>) -> io::Result<()> {
        for_both!(self, inner => inner.write_fmt(fmt))
    }

    #[inline]
    fn flush(&mut self) -> io::Result<()> {
        for_both!(self, inner => inner.flush())
    }
}

impl<L, R> WriteExt for Either<L, R>
where
    L: WriteExt + io::Write,
    R: WriteExt + io::Write,
{
    #[inline]
    fn reserve_with(&mut self, additional: usize) -> io::Result<&mut [MaybeUninit<u8>]> {
        for_both!(self, inner => inner.reserve_with(additional))
    }

    #[inline]
    #[allow(unsafe_code)] // We just dispatch over already unsafe implementations
    unsafe fn flush_len(&mut self, additional: usize) -> io::Result<()> {
        for_both!(self, inner => inner.flush_len(additional))
    }
}
