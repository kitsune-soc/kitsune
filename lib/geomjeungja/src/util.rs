use std::{fmt, ops::Deref};

#[derive(Clone)]
pub struct OpaqueDebug<T>(pub T);

impl<T> Deref for OpaqueDebug<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> fmt::Debug for OpaqueDebug<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct(std::any::type_name::<T>())
            .finish_non_exhaustive()
    }
}
