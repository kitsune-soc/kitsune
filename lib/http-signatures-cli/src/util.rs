use owo_colors::{OwoColorize, Stream};
use std::fmt::Display;

#[inline]
pub fn error_kaomoji() -> impl Display {
    "(┬┬﹏┬┬)".if_supports_color(Stream::Stdout, |text| text.red())
}

#[inline]
pub fn success_kaomoji() -> impl Display {
    "(^///^)".if_supports_color(Stream::Stdout, |text| text.green())
}
