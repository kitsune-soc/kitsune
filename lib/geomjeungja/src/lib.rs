#![doc = include_str!("../README.md")]
#![forbid(missing_docs)]

#[macro_use]
extern crate tracing;

use crate::util::OpaqueDebug;
use async_trait::async_trait;
use hickory_resolver::{
    config::{ResolverConfig, ResolverOpts},
    TokioAsyncResolver,
};
use rand::{
    distributions::{Alphanumeric, DistString},
    RngCore,
};
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use std::{convert::Infallible, future::Future, ops::Deref, sync::Arc};

mod util;

type BoxError = Box<dyn std::error::Error + Send + Sync>;
type Result<T, E = Error> = std::result::Result<T, E>;

const TOKEN_LENGTH: usize = 40;

/// Combined error type
#[derive(Debug, thiserror::Error)]
pub enum Error {
    /// The builder was incomplete
    ///
    /// The field is the name of the missing field
    #[error("Incomplete Builder: Field \"{0}\" is missing from the builder")]
    IncompleteBuilder(&'static str),

    /// The resolver returned an error
    #[error(transparent)]
    Resolve(BoxError),

    /// The verification strategy errored out
    #[error(transparent)]
    VerificationStrategy(BoxError),

    /// The domain did not have the required TXT record
    #[error("The domain did not have a TXT record matching the requirements")]
    Unverified,
}

/// DNS resolver
#[async_trait]
pub trait DnsResolver: Send + Sync {
    /// Resolve an FQDN and return its TXT records
    async fn lookup_txt(&self, fqdn: &str) -> Result<Vec<String>, BoxError>;
}

#[async_trait]
impl DnsResolver for TokioAsyncResolver {
    async fn lookup_txt(&self, fqdn: &str) -> Result<Vec<String>, BoxError> {
        let records =
            self.txt_lookup(fqdn)
                .await?
                .iter()
                .flat_map(|record| {
                    record.txt_data().iter().filter_map(|data| {
                        simdutf8::basic::from_utf8(data).ok().map(ToOwned::to_owned)
                    })
                })
                .collect();

        Ok(records)
    }
}

/// Domain verification strategy
pub trait VerificationStrategy: DeserializeOwned + Serialize {
    /// Error returned by this verification strategy
    type Error: Into<BoxError>;

    /// Verify whether the domain is valid by looking at its TXT records
    fn verify<'a>(
        &self,
        txt_records: impl Iterator<Item = &'a str> + Send,
    ) -> impl Future<Output = Result<bool, Self::Error>> + Send;
}

/// Dummy strategy that always resolves to `true`
///
/// Only useful for testing
#[derive(Default, Deserialize, Serialize)]
pub struct DummyStrategy {
    _priv: (),
}

impl VerificationStrategy for DummyStrategy {
    type Error = Infallible;

    async fn verify<'a>(
        &self,
        _txt_records: impl Iterator<Item = &'a str> + Send,
    ) -> Result<bool, Self::Error> {
        Ok(true)
    }
}

/// The de-facto default strategy
///
/// It checks whether the TXT records contain a value looking like `[key]=[value]`
#[derive(Clone, Deserialize, Serialize)]
pub struct KeyValueStrategy {
    /// Key of the entry
    pub key: String,

    /// Value of the entry
    pub value: String,
}

impl KeyValueStrategy {
    /// Create a [`KeyValueStrategy`] with a randomly generated value
    pub fn generate<R>(rng: &mut R, key: String) -> Self
    where
        R: RngCore,
    {
        Self {
            key,
            value: Alphanumeric.sample_string(rng, TOKEN_LENGTH),
        }
    }
}

impl VerificationStrategy for KeyValueStrategy {
    type Error = Infallible;

    async fn verify(
        &self,
        txt_records: impl Iterator<Item = &str> + Send,
    ) -> Result<bool, Self::Error> {
        Ok(txt_records
            .filter_map(|record| record.split_once('='))
            .any(|(key, value)| key == self.key && value == self.value))
    }
}

/// Construct the default resolver used by this library
#[must_use]
pub fn default_resolver() -> Arc<dyn DnsResolver> {
    Arc::new(TokioAsyncResolver::tokio(
        ResolverConfig::default(),
        ResolverOpts::default(),
    ))
}

/// Verifier for an arbitrary FQDN
#[derive(Clone, Debug)]
pub struct Verifier<S>
where
    S: VerificationStrategy,
{
    fqdn: String,
    strategy: S,
    resolver: OpaqueDebug<Arc<dyn DnsResolver>>,
}

impl<S> Verifier<S>
where
    S: VerificationStrategy,
{
    /// Create a new verifier
    pub fn new(mut fqdn: String, strategy: S) -> Self {
        // Since this is supposed to be a FQDN, we can just push a dot to the end of it
        // This will speed up the query to the resolver
        if !fqdn.ends_with('.') {
            fqdn.push('.');
        }

        Self {
            fqdn,
            strategy,
            resolver: OpaqueDebug(default_resolver()),
        }
    }

    /// Return the FQDN
    #[must_use]
    pub fn fqdn(&self) -> &str {
        &self.fqdn
    }

    /// Return verification strategy
    pub fn strategy(&self) -> &S {
        &self.strategy
    }

    /// Verify whether the TXT records of the FQDN pass the verification strategy
    ///
    /// Returns `Ok(())` when the check succeeded and the token is present
    #[instrument(skip_all, fields(%self.fqdn))]
    pub async fn verify(&self) -> Result<()> {
        let txt_records = self
            .resolver
            .lookup_txt(&self.fqdn)
            .await
            .map_err(Error::Resolve)?;

        let is_valid = self
            .strategy
            .verify(txt_records.iter().map(Deref::deref))
            .await
            .map_err(|err| Error::VerificationStrategy(err.into()))?;

        is_valid.then_some(()).ok_or(Error::Unverified)
    }
}

#[cfg(test)]
mod test {
    use crate::{DummyStrategy, KeyValueStrategy, Verifier};
    use rand::SeedableRng;
    use rand_xorshift::XorShiftRng;

    const RNG_SEED: [u8; 16] = *b"im breaking down";

    #[test]
    fn key_value_strategy_schema() {
        let kv_strategy =
            KeyValueStrategy::generate(&mut XorShiftRng::from_seed(RNG_SEED), "key".into());

        insta::assert_json_snapshot!(kv_strategy);
    }

    #[test]
    fn normalizes_to_fqdn() {
        let domain_verifier = Verifier::new("aumetra.xyz".into(), DummyStrategy::default());
        assert_eq!(domain_verifier.fqdn(), "aumetra.xyz.");

        let fqdn_verifier = Verifier::new("aumetra.xyz.".into(), DummyStrategy::default());
        assert_eq!(fqdn_verifier.fqdn(), "aumetra.xyz.");
    }
}
