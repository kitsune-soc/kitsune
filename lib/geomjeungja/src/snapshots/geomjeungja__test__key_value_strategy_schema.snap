---
source: lib/geomjeungja/src/lib.rs
expression: kv_strategy
---
{
  "key": "key",
  "value": "DbdoRlJiK0lLTnIyL5pciusfxinHAN3EQm1E7ITv"
}
